lib/pbench/server/api/resources/query_controllers.py

```diff
@@ -0,0 +1,273 @@
+            "size": 0,  # Don't return "hits", only aggregations
+            "aggs": {
+                "controllers": {
+                    "terms": {"field": "run.controller", "order": [{"runs": "desc"}]},
+                    "aggs": {"runs": {"max": {"field": "run.start"}}},
+                }
+            },
+        }
+
+        # TODO: the big assumption here involves the index version we're
+        # referencing. While it's possible the root index name could change
+        # at some point, advancing the version is going to happen more
+        # often.
+        #
+        # So ... do we need to change the code here each time? The old
+        # dashboard config is hardcoded for each version, so one option is
+        # to add a version for each index to the pbench-server.cfg file.
+        #
+        # When we have a full live Python server, persisting the template
+        # information in Redis or postgreSQL probably makes sense; maybe
+        # we just hardcode until then?
+        uri_fragment = QueryControllers._genMonthUriRange(
+            prefix, ".v6.run-data.", start, end
+        )
+
+        uri = f"{self.app.config['ES_URL']}/{uri_fragment}/_search"
+        logger.info("POSTING query to {}: {}", uri, payload)
+        try:
+            # query Elasticsearch
+            es_response = requests.post(
+                uri,
+                params={"ignore_unavailable": "true"},
+                headers={
+                    "Content-Type": "application/json",
+                    "Accept": "application/json",
+                },
+                json=payload,
+            )
+        except requests.exceptions.HTTPError as e:
+            logger.exception("HTTP error {} from Elasticsearch post request", e)
+            abort(500, message=f"HTTP error {e} from Elasticsearch")
+        except requests.exceptions.ConnectionError:
+            logger.exception("Connection refused during the Elasticsearch post request")
+            abort(502, message="Network problem, could not post to Elasticsearch")
+        except requests.exceptions.Timeout:
+            logger.exception(
+                "Connection timed out during the Elasticsearch post request"
+            )
+            abort(
+                504, message="Connection timed out, could not post to Elasticsearch",
+            )
+        except requests.exceptions.InvalidURL:
+            logger.exception(
+                "Invalid url {} during the Elasticsearch post request", uri
+            )
+            abort(
+                500,
+                message=f"Invalid Elasticsearch url {uri}, could not complete the post request",
+            )
+        except Exception as e:
+            logger.exception(
+                "Exception {!r} occurred during the Elasticsearch post request", e
+            )
+            abort(500, message=f"Could not post to Elasticsearch: {e!r}")
+        else:
+            es_json = es_response.json()
```

Comment:
It seems like we're going to need to wrap this entire structure with a `try/except` block, not just for lines 260 - 267 below.

Reply:
Don't we need to verify the response code as well? If we receive a 400 error, wouldn't we want to deal with that properly instead of 500 error logged?

Reply:
I was assuming that we'd never get a non-200 return without an exception. Is that not true?

Plus, aside from 408, which ought to be a Timeout exception, I'm not sure how we could handle an ES 400 error except as a 500 ... either we failed to properly validate our parameters from the client, or we fabricated the ES URI or payload incorrectly. Do you have any specific recommendations?

And, yeah; I should have put the `es_response.json()` inside the `try` in case there's no JSON response at all.

Reply:
> Don't we need to verify the response code as well? If we receive a 400 error, wouldn't we want to deal with that properly instead of 500 error logged?

This turns out to be one of those unexpectedly prescient questions, Pete. I investigated the `requests` documentation to see if it's possible to return an error without raising an exception.

And, um, it turns out that `requests` __never__ raises an exception. Unless you tell it to. I verified this by breaking my ES query and tickling the modified server to get an ES 400 response. It went right through my `else:`.

Turns out there's a `raise_for_status` method on the response object that'll raise the appropriate exception if the response isn't a 200 range success `status_code`. I've added that following my call as well as in the pass-through ES and GQL queries!